### PR TITLE
[Release] - Bake cancellation token into Aff<T> (#1057) 

### PR DESCRIPTION
**Pull request by bmazzarol**
@louthy had a play with some options based on #1057.

I tried the simplest thing first, just add CancellationToken to the thunk in Aff<T>.

TBH I think it works really well, let me know what you think. 

The rest is just a copy of the commit message.

The current version of Aff<T> cannot be cancelled.

This simply bakes a cancellation token into the thunk. This makes an Aff<T> very similar to Async in F#, a cold task with automatic cancellation support.

Note: Transformers that go from a basic async type to Aff currently do not propagate cancellation tokens. Such as TryAsync and OptionAsync

BREAKING CHANGE: 

* Fork now returns an Eff that can be used to cancel and rejoin the task, the same as Aff<RT,T>
* Rx extensions Fold and Consumer now do not take cancellation tokens. They are provided by Aff

### CODE REVIEW

**Review by louthy — CHANGES_REQUESTED** [review]
Firstly, welcome to the pain of changing a small thing in this library :P 

I think this is close, but the `Fork` behaviour could do with some work to be more like `Aff<RT, A>.Fork`.

There's also an argument for wrapping up `CancellationToken` in a new type that means it could be extended later (if some other sensible default behaviour should flow through).  I'm not entirely convinced it's necessary just yet, but maybe work considering.

File: LanguageExt.Core/Effects/Aff/AffPure.cs [hunk 1]
```diff
@@ -80,10 +89,20 @@ public Aff<A> Memo()
...
+            {
+                var cts = new CancellationTokenSource();
+                var task = Task.Run(async () => await t(cts.Token), cts.Token);
+                return Eff(() =>
+                {
+                     cts.Cancel();
+                     task.Result.ThrowIfFail();
+                     return unit;
+                });
+            });
```
[the remark this answers is not in the document]
Comment (bmazzarol) [reply LanguageExt.Core/Effects/Aff/AffPure.cs:117]: @louthy I have added support this way,

Created a [localCancel()](https://github.com/louthy/language-ext/pull/1068/files#diff-c00aee082933c2e24db392afd10f878b31eef162769220bc974c5007b4fdc82fR237-R243) function that allows access to the current cancellation token threading through Aff<T>.

Then added an [optional cancellation token parameter](https://github.com/louthy/language-ext/pull/1068/files#diff-c00aee082933c2e24db392afd10f878b31eef162769220bc974c5007b4fdc82fR181) to Fork/fork which will be linked into the cancellation token source.

Now the parent Aff can have its provided cancellation token cancel any forked processes.

Here is an example [here](https://github.com/louthy/language-ext/pull/1068/files#diff-d1ade6418fb5851423317a4f48bd7fa3411f0916bcd7004a73191050c9a3a8a5R122-R130)

Let me know what you think.

**Comment by bmazzarol** [comment]
@louthy have look through and let me know if you are happy with the fork changes.

As for wrapping cancellation token up, I see no issue. What should that type be called? DefaultRuntime? 

Do you want a variant of the constructor functions for the new type as well as just cancellation token?

**Comment by bmazzarol** [comment]
@louthy I had a go at the other async types as well, they are not amenable to threading cancellation through in the same way (as I'm sure you were already aware). I had to have a go though.

I got close with TryAsync, the issue was that its not possible to implicitly convert Func to a custom type. Therefore any change to the delegate was breaking.

So as far as the scope of this goes I don't have any more I want to add here. (Unless you want me to convert the the cancellation token to a struct type, which I'm more than happy to do just say the word)

If you think its not worth the effort to merge this in let me know and I will close it off.

I will plead my case in favour of it,

* Removes the requirement to have a cancellation tokens in the parameter list. (Literally thousands of lines of just CancellationToken cancellationToken = default in my workplace)
* Allows for parity with the forking behaviour between Aff<T> and Aff<RT,T>
* Allows for parity around all retry/repeat methods between Aff<RT,T> and Aff<T>
* Provides an additional incentive to use Aff in place of TryAsync or (shudder, shudder) Task

**Comment by louthy** [comment]
Hold off on the work on this.  I'm starting a refactor for version 5, the internals of `Aff` and `Eff` will become a DSL - so most of what you're working on will likely be different (internally).  I'll work in the concept of a cancellation-token into the inner DSL (which will be shared between `Aff<A>` and `Aff<RT, A>`

**Comment by bmazzarol** [comment]
Sure thanks for the detail, will use Aff<RT, T> for all cases requiring cancellation.

Looking forward to seeing what you come up with  in V5.

**Pull request closed without merging**